src/java/org/apache/cassandra/metrics/ThreadLocalMeter.java

```diff
@@ -0,0 +1,372 @@
+        allMeters.add(new WeakReference<>(this));
+        ThreadLocalMetrics.destroyWhenUnreachable(this, new MeterCleaner(countMetricId, uncountedMetricId, rateGroupId));
+        ReflectionUtils.setFieldToNull(this, "count"); // to reduce metrics memory footprint
+    }
+
+    private static class MeterCleaner implements ThreadLocalMetrics.MetricCleaner
+    {
+        private final int countMetricId;
+        private final int uncountedMetricId;
+        private final int rateGroupId;
+
+        private MeterCleaner(int countMetricId, int uncountedMetricId, int rateGroupId)
+        {
+            this.countMetricId = countMetricId;
+            this.uncountedMetricId = uncountedMetricId;
+            this.rateGroupId = rateGroupId;
+        }
+
+        @Override
+        public void clean()
+        {
+            recycleRateGroupId(rateGroupId);
+            ThreadLocalMetrics.recycleMetricId(countMetricId);
+            ThreadLocalMetrics.recycleMetricId(uncountedMetricId);
+        }
+
+        private static void recycleRateGroupId(int rateGroupId)
+        {
+            synchronized (freeRateGroupIdSetGuard)
+            {
+                freeRateGroupIdSet.set(rateGroupId);
+            }
+        }
+    }
+
+    /**
+     * Mark the occurrence of an event.
+     */
+    public void mark()
+    {
+        mark(1);
+    }
+
+    /**
+     * Mark the occurrence of a given number of events.
+     *
+     * @param n the number of events
+     */
+    public void mark(long n)
+    {
+        // we retrieve the context once here to reduce thread local lookup overheads
+        ThreadLocalMetrics context = ThreadLocalMetrics.get();
+        context.addNonStatic(countMetricId, n);
+        context.addNonStatic(uncountedMetricId, n);
+    }
+
+    @Override
+    public long getCount()
+    {
+        return ThreadLocalMetrics.getCount(countMetricId);
+    }
+
+    @Override
+    public double getFifteenMinuteRate()
+    {
+        return getRatePerSecond(getRateValue(rateGroupId + M15_RATE_OFFSET));
+    }
+
+    @Override
+    public double getFiveMinuteRate()
+    {
+        return getRatePerSecond(getRateValue(rateGroupId + M5_RATE_OFFSET));
+    }
+
+    @Override
+    public double getOneMinuteRate()
+    {
+        return getRatePerSecond(getRateValue(rateGroupId + M1_RATE_OFFSET));
+    }
+
+    @Override
+    public double getMeanRate()
+    {
+        long count = getCount();
+        if (count == 0)
+            return 0.0;
+        else
+        {
+            final double elapsed = clock.now() - startTime;
+            return count / elapsed * TimeUnit.SECONDS.toNanos(1);
+        }
+    }
+
+    @VisibleForTesting
+    static void tickAll()
+    {
+            List<WeakReference<ThreadLocalMeter>> emptyRefsToRemove = null;
+            for (WeakReference<ThreadLocalMeter> threadLocalMeterRef : allMeters)
+            {
+                ThreadLocalMeter meter = threadLocalMeterRef.get();
+                if (meter != null)
```

Comment:
suggestion: either pull the first branch up to the same line as the if, or add the braces to match the else branch. I find it easier to read when braces and new lines are balanced, so either:

```
if (meter != null) meter.tickIfNessesary();
else
{
...
}
```

or

```
if (meter != null) 
{
    meter.tickIfNessesary();
}
else
{
...
}
```